### PR TITLE
Don't use ivars directly

### PR DESCRIPTION
**Pull request by thorncp**
Didn't focus on the strings yet because this diff is already a little difficult to follow. Planning to do strings in a follow-up, and perhaps some other refactors related to the things touched here..

### CODE REVIEW

File: lib/app.rb [hunk 1]
```diff
@@ -10,145 +10,139 @@ class App
...
+      if no_current_project?
         case command
         when "a"
-          @projects = [] if @projects.nil?
-          @output_stream.puts("\e[0;3mEnter a project name:\e[0m")
-          name = @input_stream.gets.chomp
-          @projects << {name => []}
-          @output_stream.puts("\e[38;5;40mCreated project:\e[0m '#{name}'\n\n")
+          capture_project_name
+          projects << {project_name => []}
```
Comment (baccigalupi) [inline lib/app.rb]: In this case you got rid of the ivar, but didn't extract it into a named method. More than moving the ivars into accessors, I am thinking of the ivars and what they represent. Here the ivar was part of the data layer and I would have taken the whole line to make into a command.

File: lib/app.rb [hunk 2]
```diff
@@ -10,145 +10,139 @@ class App
...
+          output_stream.puts("\e[38;5;40mCreated project:\e[0m '#{project_name}'\n\n")
```
Comment (baccigalupi) [inline lib/app.rb]: I'm feeling the same way about `@output_stream` extraction. Seems like `@output_stream.puts` to `print` would be a better path through this morass.

File: lib/app.rb [hunk 3]
```diff
@@ -10,145 +10,139 @@ class App
...
         when "ls"
-          @output_stream.puts("\e[38;5;40mListing projects:\e[0m\n")
-          if !@projects.nil? && !@projects.empty?
-            @projects.each do |project|
-              @output_stream.puts("  #{project.keys.first}\n")
+          output_stream.puts("\e[38;5;40mListing projects:\e[0m\n")
+          if projects?
```
Comment (baccigalupi) [inline lib/app.rb:40]: Awesome!

File: lib/app.rb [hunk 4]
```diff
@@ -10,145 +10,139 @@ class App
...
+            projects.each do |project|
```
Comment (baccigalupi) [inline lib/app.rb:41]: And also this! Love how wrapping up the data model makes things more understandable.

File: lib/app.rb [hunk 5]
```diff
@@ -10,145 +10,139 @@ class App
...
         when "d"
-          if @projects && (@projects.size > 0)
-            @output_stream.puts("\e[0;35mEnter a project name:\e[0m")
+          if projects?
+            output_stream.puts("\e[0;35mEnter a project name:\e[0m")
             capture_project_name
-            @deleted = @projects.delete_if { |project| project.keys.first == project_name.strip }.empty?
-            if @deleted
-              @output_stream.puts "\e[38;5;40mDeleting project:\e[0m '#{project_name.strip}'\n\n"
+            deleted = projects.delete_if { |project| project.keys.first == project_name.strip }.empty?
```
Comment (baccigalupi) [inline lib/app.rb]: This whole thing seems like it should be extracted into a data model method.

File: lib/app.rb [hunk 6]
```diff
@@ -10,145 +10,139 @@ class App
...
+            output_stream.puts("\e[40;38;5;214mProject doesn't exist:\e[0m '#{project_name}'\n\n")
           end
         end
       else
         case command
         when "a"
           capture_task_name
-          @current_project.values.first << task_name
-          @output_stream.puts("\e[38;5;40mCreated task:\e[0m '#{task_name}'\n\n")
+          current_project.values.first << task_name
```
Comment (baccigalupi) [inline lib/app.rb]: And this.

**Review by baccigalupi — CHANGES_REQUESTED** [review]
While I think noticing the colors of ivars as a smell is really good, looking only at wrapping up these into ivars is sending this not in the right direction.

I noted the `output.puts` repetition that seems like the better extraction. But also you left behind a lot of model logic instead of grabbing each ivar and figuring out in context what to extract.

I think you should take a pass at being a bit less strict and literal about following the ivar color to see what you get to.

Also, how should me change the instruction to reflect this?

**Comment by thorncp** [comment]
> I think you should take a pass at being a bit less strict and literal about following the ivar color to see what you get to.
> 
> Also, how should me change the instruction to reflect this?

I don't know if the instructions need to change. What you lined out is along the same lines as my intended destination. I got fixated on removing the ivars along the way and the diff started getting difficult to follow, so I opened a PR and planned to make some follow up refactors after. I think what I should have done to reduce the diff was pick out specific ivars and refactor them, then follow-up with other ivar refactors.

File: lib/app.rb [hunk 7]
```diff
@@ -10,145 +11,137 @@ class App
   def initialize(output_stream, input_stream)
     @output_stream = output_stream
     @input_stream = input_stream
+    @projects = []
```
Comment (baccigalupi) [inline lib/app.rb:14]: This is great! Making sure the data is there as an array from the beginning solves so many possible issues.

**Review by baccigalupi — APPROVED** [review]
This looks like a better step towards dealing with ivar coloring! Can't wait to see the next step

**Pull request merged**